homeassistant/components/mqtt/quality_scale.yaml

```
@@ -29,7 +29,7 @@ rules:
       MQTT broker, this happens during integration setup,
       and only one config entry is allowed.
   test-before-configure: done
-  test-before-setup: done
+  test-before-setup: todo
```

@jbouwh I couldn't see how this was implemented for `mqtt`.
It doesn't raise any of the documented exceptions, and doesn't call `async_config_entry_first_refresh`

I've marked it back as `todo`

If it's a problem with the validation, then I'm happy to tweak accordingly

---

Right, I'll have a look at this. As it is now we leave the entry setup early as the mqtt set up as backend process. Integrations should call `await mqtt.async_wait_for_mqtt_client(hass)` before they subscribe any topics to MQTT or start publishing.

---

I did check if a test can be done before setup, but it seems we choose to early exit the entry as it can take some time for the client to connect. Waiting for the client would increase the overall setup time. So I feel not such a validation is very helpfull for MQTT.
@emontnemery what d you think?

---

I don't think we should wait for the client. It's fine to not be connected to the client initially, other integrations are still able to set up subscriptions etc.

---

Are you happy with this suggestion?

```suggestion
  test-before-setup:
    status: exempt
    comment: >
      We choose to early exit the entry as it can take some time for the client 
      to connect. Waiting for the client would increase the overall setup time.
```

---

Looks good to me